### ISSUE TITLE
a recursive call is not 100% reliably tail recursive, perhaps provoked by Arch

### ISSUE THREAD

**Issue by callegar** [bcpierce00/unison#1006]
Using unison version 2.53.4 (ocaml 5.1.0), I need to sync with an host running unison version 2.51.5 (ocaml 4.13.1).

It is unclear to me if this combination is supported at all. Unison does not complain, though.

The changes are collected just fine.

When I tell unison to propagate the changes it starts, but at about 50% it crashes, with a segmentation fault. Even if the version combination above is weird, IMHO unison should at most refuse to run, not segfault.

Unfortunately I cannot install a more recent unison on the target machine to test with it.

Turning on debugging, shows `[files] Creating directory <long path>` as the last action before the crash. From the path, unison is trying to transfer a piece of a `.git` directory (in fact an object directory).

Because of how git objects are named, I suspect a rather long path.

**Comment by callegar** [comment]
Maybe same as https://github.com/bcpierce00/unison/issues/990

Machine on which unison is invoked is also an arch derivative.

**Comment by tleedjarv** [comment]
Yes, it appears to be the same issue as https://github.com/bcpierce00/unison/issues/990#issuecomment-1965712014

@gdt, is it possible to move comments starting from https://github.com/bcpierce00/unison/issues/990#issuecomment-1965712014 out of #990 and into this ticket here? It seems that the original #990 is a different issue but the linked comment and this ticket (#1006) are the same issue. If this is not possible then we'll just have to keep this slight mix-up in mind and treat #1006 as the main ticket for this issue going forward.

**Comment by norbusan** [comment]
To add one point here: I had these crashes with client and server on the same version of ocaml, not like the OP above.

**Comment by tleedjarv** [comment]
My current understanding is that this is a problem with OCaml 5.x (or maybe specifically 5.1.x). OCaml 5.x series is currently still considered experimental.

Unison 2.53.4 compiled with OCaml 4.x does not seem to crash like this.

**Comment by tleedjarv** [comment]
> Using unison version 2.53.4 (ocaml 5.1.0), I need to sync with an host running unison version 2.51.5 (ocaml 4.13.1).
> 
> It is unclear to me if this combination is supported at all. Unison does not complain, though.

I think this combination should work and I think the segfault is not related to this version mismatch. It may well be, but it is difficult to say without a proper backtrace.

@callegar instead of the version built with OCaml 5.1.0, could you grab the build from Github releases page (which is built with 4.14), or build yourself with OCaml 4.12 or 4.13 or 4.14 and try again? That _should_ work without any issues with version 2.51.5 (ocaml 4.13.1). This combination is expected to work for these OCaml versions, but all bets are off for 5.x series.

If you can't try a different version and if you still get a segfault then could you get a backtrace from the crash?

**Comment by gdt** [comment]
I think this should be retitled to "(probably) ocaml 5 miscompilation leads to segfault", at least for now, until we understand what is happening better.

**Comment by gdt** [comment]
No, I can't move omments.

**Comment by callegar** [comment]
> @callegar instead of the version built with OCaml 5.1.0, could you grab the build from Github releases page

I can, but it won't help me check it it fixes the issue, since the directory that was giving me problems has already been (painfully) synchronized manually with rsync.

 > I think this should be retitled to "(probably) ocaml 5 miscompilation leads to segfault", at least for now, until we understand what is happening better.

So you suspect that the ocaml 5 compiler does not play well with unison? Isn't issue https://github.com/bcpierce00/unison/issues/990 more likely, given that I was synchronizing a directory that had nested inside a git repo with a lot of loose objects leading to a very large number of files?

**Comment by gdt** [comment]
That is Toivo's suspicion as I understand it.    My basic concept is to have one bug open per theorized defect, at least for now.

If you or someone wants to reproduce problems, and preferably write a script that will reproduce them, we can test and look into it.   I realize you just want to get past this -- and I would too, but not creating a repro recipe/script leads to "zero people currently have this problem".   For now we have no examples of this segfault with mainstream ocaml, just new/experiental ocaml.

**Comment by callegar** [comment]
>  I realize you just want to get past this

More than this, the matter is that it is quite hard to pack a test case: when I encountered the bug it was on a largish tree, with a git repo inside with a lot of loose objects and data that I cannot freely share.

What is really needed is a "synthetic" example that can be generated on demand, but that is not easy to do.

In the meantime, if you suspect that ocaml 5 may break unison, probably it would be good to put a notice about this in evidence for distributions and packagers: my ocaml/unison combination comes from a linux distro and I suspect many of them do not consider ocaml 5 experimental and in absence of any strong motivation to keep it are dropping version 4 (fedora and derivatives, arch and derivatives, etc.)

**Comment by gdt** [comment]
The upstream ocaml webpage says
```
OCaml 5.1.0 is still a relatively experimental release compared to the OCaml 4.14 branch. In particular
```

so I don't think it's unison's place to have to warn about that.  Given that statement, packagers should not have the main ocaml package be the main one.

**Comment by tleedjarv** [comment]
I do suspect these are different issues and there is at least one other user whose experience seems to confirm that. It is difficult to say without a crash trace or other information about your setup.

If you still have the binary you ran when it crashed, could you attach it here so I can perhaps investigate further. Perhaps you still have the core dump from the crash lying around somewhere? If yes, please attach it here.

If you can then please post some information about your configuration. We don't need the common preferences such as root, path, ignore, etc. But it would be good to know if you're using some lesser-used preferences such as atomic, maxthreads or stream. Could you estimate the number of updates that led to the crash? Finally, please clarify if you were using the text UI or the GUI.

I do not think it is #990 because there are currently no known stack overflow bugs and I test with relatively small stacks (< 100KB whereas default on most systems is 8MB or even 10MB) and relatively large replicas (100 000+ updates, sometimes 1 000 000+ updates).

**Comment by callegar** [comment]
Just noticed that the unison version provided in arch linux carries this downstream patch:

https://gitlab.archlinux.org/archlinux/packaging/packages/unison/-/blob/main/unison-large.patch?ref_type=heads

That is a patch meant to fix https://github.com/bcpierce00/unison/issues/48

Is that still needed at all? Can it be the cause of the misbehavior?

**Comment by tleedjarv** [comment]
No, that patch is ancient. It can't even apply on the current code.

**Comment by FSMaxB** [comment]
> Can it be the cause of the misbehavior?

I built the package without the patch and it is still crashing. So if it is the cause, it's not all of it.

**Comment by tleedjarv** [comment]
How do you know it is still crashing? I thought there was no reproducer. If you can reproduce the crash, can you extract a stack trace or a core dump from it? Can you pinpoint which conditions are required for the crash?

**Comment by FSMaxB** [comment]
> How do you know it is still crashing?

I'm syncing the day's changes at the end of my workday and on *most* days it triggers the crash :sweat_smile:. If that happens, I need to run unison multiple times until the crash to incrementally get all files synced.

> can you extract a stack trace or a core dump from it?

Apparently systemd has captured a coredump for me. This is the backtrace [unison-gui-backtrace.txt](https://github.com/bcpierce00/unison/files/14770440/unison-gui-backtrace.txt)

Note that this is the GUI, I haven't actually tried it with the CLI yet because the CLI isn't really practical for larger numbers of files.

**Comment by tleedjarv** [comment]
I see, I mixed up the reporters :)

Thank you for the backtrace, this has been most helpful. The trace looks very much the same as the one @norbusan provided, so it most likely is the same issue and for now I'm still very much convinced that this is a OCaml 5.x issue (not to say that the bug still isn't in Unison code, just that it is not triggered by OCaml 4.x).

What is the exact Unison version you are using? If you can trigger the crash so frequently then my first advice is to try and compile with OCaml 4.x or download from the GH releases page and then see if the crashes go away.

**Comment by FSMaxB** [comment]
> What is the exact Unison version you are using? 

It's the archlinux version, so `2.53.4` compiled with OCaml `5.1.1`.

> If you can trigger the crash so frequently then my first advice is to try and compile with OCaml 4.x or download from the GH releases page and then see if the crashes go away.

Yeah, makes sense. I was just not doing it by now because things have been kind of busy lately and I only ever remember about the crashes when they happen at the end of the day when I have other places to be and things to do :laughing:.

But given that I have my own `PKGBUILD` now anyways to test removing the archlinux downstream patch I can just as well download the release binaries from github in there and try that.

**Comment by tleedjarv** [comment]
Yes, please do, that would be very helpful.

Another thing, which presumably also comes from arch build scripts, is that both your and @norbusan's traces show that ocaml itself is stripped of debug symbols. _If_ compiling with 5.1.1, is it possible to keep the debugging symbols in the compiler/compiler stdlib?

**Comment by tleedjarv** [comment]
I have been trying to reproduce the issue myself but have been unsuccessful so far. @callegar @FSMaxB @norbusan if you have a binary that you know would crash, please attach it here so I can take a look at it.

Right now it seems there are three unique reporters reporting the same exact issue (but not 100% sure because we didn't get a backtrace from one reporter). There are a couple of things common between the reporters:
 * a binary compiled with OCaml 5.x
 * using arch or a derivative
 * all using the GUI? (perhaps not)
 
Seeing as arch is the common theme here, I looked at its build recipe for ocaml-5 but couldn't really see anything suspicious. They do enable frame pointers, though. I also tried my own build with OCaml 5.1.1 with frame pointers enabled. Aside from making things run slower, still could not reproduce a crash.

We really need your help to find out what's going on here, so I ask all the reporters for following:
 * if you can, switch to a binary from https://github.com/bcpierce00/unison/releases (built with OCaml 4.14.x) (as I understand the `unison-upstream-bin` AUR can be used for this purpose) OR build yourself with OCaml 4.x; please report if you still see crashes. @norbusan has already reported that the crashes stopped.
 * if you have established that the crashes stopped with the bullet above then please try a binary from https://github.com/tleedjarv/unison/actions/runs/8481008964 which is 2.53.4 built with OCaml 5.1.1 the same way as the released binaries are built.

Additionally, please provide as much information as possible about your setup:
 * what is the approximate size of the replica or number of updates that you see causing a crash? More precisely, if you know, about how many updates into the sync when it crashes?
 * are you using any lesser-used preferences such as atomic, maxthreads or stream?
 * are you seeing crashes on a local sync or is it always a remote sync?
 * CPU arch?
 
Since I've been unable to reproduce the crash with binaries built with OCaml 5.x then I can only suspect that there is either something in your replicas/configuration that is a precondition for the crash or that there is something in the arch build process. It's not impossible that the arch build process may make the compiler produce different code (mainly due to the runtime itself being impacted); or a linker produce a different executable; or the problem could be in a linked library, such as lablgtk/GTK.

**Comment by tleedjarv** [comment]
So I went ahead and installed Arch to test this.

This is definitely an Arch issue. You may have to report this to Arch.

At first I thought this is a stack overflow but it might not be (see below). Additionally, the Arch built binary is _much-much_ slower than other builds (like the ones downloaded from the releases page or the ones built by GH CI, whatever the compiler version) and clearly uses much more CPU (as evidenced by the fan taking off like crazy).

My current theory is that it may be caused by a combination of the Arch build process (the terrible slowness), and the dynamic linker and/or glibc of Arch (the stack growth). Either way, the exact same binaries behave differently on different systems and they **do not crash on a non-Arch system**.

How I tested on Arch:
 * Limit the stack to 144 KB (`ulimit -s 144`) (because at first I thought this was related to a stack overflow, and this seems to be the smallest stack that GTK will run with).
 * Sync 3689 updates (locally) - a random largish number.
 * OCaml 5 builds (non-Arch, produced by GH CI) sync successfully.
 * OCaml 4.14.1 build crashes after about 3370 updates synced; but succeeds with a 160 KB stack.
 * Arch build crashes after only 630 updates synced. **This seems to be a hard limit and unrelated to the stack size because it always crashes at 630 updates synced, even with the default 8 MB stack.**

Why do I think this is an Arch problem?

Because I can sync with the exact same binaries, even the Arch-built one, with the exact same tiny stack size (144 KB; and less for TUI) on my own machine a basically unlimited number of updates (I tested with up to 400 000 updates, except with the Arch-built binary because it is _extremely_ slow). Peeking at the trace of the running process, there is no stack growth whatsoever. This applies to both the GUI and TUI.

Why do I think the dynamic linker and/or glibc of Arch is (partly) to blame?

Because the exact same binaries that exhibit no stack growth whatsoever on a different system suddenly start doing so on an Arch installation. This also happens with the TUI binary which has no other dependencies. Something in the Arch dynamic linker/glibc probably breaks the tail calls and causes them to behave like regular recursive calls.

And the Arch build process?

Even though the Arch-built binary does not exhibit any unwanted stack growth on a non-Arch system, it is still extremely slow compared to binaries built on a non-Arch system and clearly abuses the CPU.

**Comment by gdt** [comment]
leaving this open so we don't get dups, but I conclude, for now, always subject to revision with more evidence, that this is not a unison bug.

**Comment by callegar** [comment]
Thanks for the extensive testing! I did not notice the slovenliness and the extra CPU load not having a term of comparison, would be great to know what the CPU is doing, particularly after the alarming news that have recently arrived from other codes.

Has this information been passed to the arch developers?  I had originally opened https://gitlab.archlinux.org/archlinux/packaging/packages/unison/-/issues/?sort=created_date&state=closed&first_page_size=20, but it was quickly closed.

**Comment by tleedjarv** [comment]
Alas, I spoke too soon...

I discovered some differences between local sync and remote sync. There may still be a Unison bug involved here but it certainly acts differently on an Arch system, regardless. I will get back shortly with my updated findings.

**Comment by callegar** [comment]
In my case, the attempt was run with the gui, via ssh. Hard to reproduce now, because rarely I get to a situation where I sync having more than 600 files out of sync. Looks like a similar situation can be created with a synthetic test case now, though.

**Comment by tleedjarv** [comment]
> particularly after the alarming news that have recently arrived from other codes.

That had crossed my mind also but this same slowness is apparent in the TUI-only binary which does not have any dependencies, so does not link with liblzma in any way.

> Has this information been passed to the arch developers?

No, and I don't plan to do that either.

> In my case, the attempt was run with the gui, via ssh. Hard to reproduce now, because rarely I get to a situation where I sync having more than 600 files out of sync. Looks like a similar situation can be created with a synthetic test case now, though.

Yes, it seems to be directly related to the number of updates to sync but at the same time not related to the stack size limit, which is really weird.

**Comment by callegar** [comment]
Most important: has the existence of the issue spotted in arch been excluded in other distros using OCaml 5.x?

I am trying to collect enough data to make the report to arch myself, then...

**Comment by tleedjarv** [comment]
I actually need to retest myself because I think I may have mixed up some binaries here. I will get back with more accurate information soon. The same conclusions may no longer hold.

**Comment by FSMaxB** [comment]
> If compiling with 5.1.1, is it possible to keep the debugging symbols in the compiler/compiler stdlib?

Not sure if I succeeded in that, but here is another stacktrace: 
[unison-gui-ocaml-debug-backtrace.txt](https://github.com/bcpierce00/unison/files/14814925/unison-gui-ocaml-debug-backtrace.txt)

I just tried to reproduce without the GUI and wasn't able, but that doesn't necessarily mean anything because the GUI was previously run, making some progress before crashing so it might just have been brought below the threshold where it crashes.

I will try with upstream release binaries next, but testing is slow because I only get to test every one or two workdays.


> what is the approximate size of the replica or number of updates that you see causing a crash? More precisely, if you know, about how many updates into the sync when it crashes?

The replica is 511GiB with 1839500 files. But only a small portion of that changes usually. I don't know how far into the sync it crashes. But I think the amount of changed items is at least several thousands when it happens.

> are you using any lesser-used preferences such as atomic, maxthreads or stream?

The only preferences I'm using are `root`, `watch = false` and `ignore`.

> are you seeing crashes on a local sync or is it always a remote sync?

Haven't tried a local sync, what crashes is a remote sync.

> CPU arch?

x86_64


I also recently removed all the archive files and tried again. This didn't fix anything, so it is definitely not a corrupted archive file.

**Comment by FSMaxB** [comment]
Ok, it also crashes when syncing locally!

And I think I found a way to reproduce it locally. When I take our frontent project, jump between versions that are far apart and run `yarn install`, then try to sync, it crashes.

Still need to find a step-by-step reproducible way that doesn't include proprietary data. One thing I'll try once I find the time (probably not today) is to maybe switch between versions in the linux kernel source tree, that might trigger it.

**Comment by tleedjarv** [comment]
I have to completely retract my previous comment (I have now deleted it because it has become misleading).
My previous conclusions were off due to a combination of testing too many variants, some missing notes and not paying enough attention.

I've tested with the following binaries (all of Unison 2.53.4):
* From the GitHub releases page, built with OCaml 4.14.1 (Ubuntu)
* Built by GitHub CI (Ubuntu) with OCaml 5.1.1
* Built by GitHub CI (Ubuntu) with OCaml 5.1.1 with frame pointers enabled
* Built by GitHub CI (Ubuntu) with OCaml 5.2.0~beta1
* Built by GitHub CI (Ubuntu) with OCaml 5.1.0
* From Arch

I tested on a freshly installed Arch system and on openSUSE.

The results:
Binary | TUI/GUI | local/remote sync | on Arch | on non-Arch | stack size limit | Comments
-- | -- | -- | -- | -- | -- | --
GH OCaml 4.14.1 | TUI | local | **stack overflow** | **stack overflow** | 80 |  
GH OCaml 4.14.1 | GUI | local | **stack overflow** | **stack overflow** | 144 |  
GH OCaml 5.1.1 | TUI | local | OK, but see *** | OK, but see *** | 80 |  
GH OCaml 5.1.1 | GUI | local | OK, but see *** | OK, but see *** | 144 |  
GH OCaml 5.1.1 FP | TUI | local | OK, but see *** | OK, but see *** | 80 |  
GH OCaml 5.1.1 FP | GUI | local | OK, but see *** | OK, but see *** | 144 |  
GH OCaml 5.2.0 beta | TUI | local |  | OK, but see *** | 80 |
GH OCaml 5.2.0 beta | GUI | local | OK, but see *** | OK, but see *** | 144 |
GH OCaml 5.1.0 | TUI | local |  | OK, but see *** | 80 | extremely slow
GH OCaml 5.1.0 | GUI | local | OK, but see *** | OK, but see *** | 144 | extremely slow
Arch OCaml 5.1.0 | TUI | local | OK, but see *** | OK, but see *** | 80 | extremely slow
Arch OCaml 5.1.0 | GUI | local | **stack overflow** | **stack overflow** | 144 | extremely slow; always syncs exact number of updates before crashing (around 620 - 630)
Arch OCaml 5.1.0 | GUI | local | **stack overflow** | **stack overflow** | 8192 | extremely slow; always syncs exact number of updates before crashing (around 620 - 630)
GH OCaml 4.14.1 | TUI | remote | OK | OK | 80 |  
GH OCaml 4.14.1 | GUI | remote | OK | OK | 144 |  
GH OCaml 5.1.1 | TUI | remote | OK | OK | 80 |  
GH OCaml 5.1.1 | GUI | remote | OK | OK | 144 |  
GH OCaml 5.1.1 FP | TUI | remote | OK | OK | 80 |  
GH OCaml 5.1.1 FP | GUI | remote | OK | OK | 144 |  
GH OCaml 5.2.0 beta | TUI | remote |  | OK | 80 |
GH OCaml 5.2.0 beta | GUI | remote | OK | OK | 144 |
GH OCaml 5.1.0 | TUI | remote |  | OK | 80 | not slow? against a non-Arch server
GH OCaml 5.1.0 | GUI | remote |  | OK | 144 | extremely slow
Arch OCaml 5.1.0 | TUI | remote | OK | OK | 80 | not slow? against a non-Arch server
Arch OCaml 5.1.0 | GUI | remote | OK | OK | 144 | extremely slow

"OK" means no crash. Simply "OK" without additional comments means no unwanted stack growth whatsoever.

*** I see something weird that I can't explain at the moment. Many of the OCaml 5.x builds run the local sync correctly without crashing but when I extract the trace of the running process with gdb I can see a large number of frames. Large enough that it should easily crash with the limited stack size. Yet, no crashes... Can anyone explain what is going on here?

New conclusions from this:
* Just _running_ the binary on Arch or a non-Arch system does not make a difference. This was incorrect in my previous comment.
* How the binary was _built_ does make a difference. OCaml 5.1.0 is definitely one cause but the Arch-built binary additionally gets different results. The takeaway is that **don't compile with OCaml 5.1.0**. But 5.1.1 seems ok.
* Local sync code path does have some weirdness and may be a Unison bug but I don't understand why it sometimes crashes and sometimes not. **I have not been able to make the remote sync crash.**

**Comment by FSMaxB** [comment]
Update: I can only reproduce the crash with `unison-gui`, not with the CLI!

With my frontend-code test-case, the GUI crashes every time. But I was able to sync with the CLI no problem.

**Comment by FSMaxB** [comment]
> The takeaway is that don't compile with OCaml 5.1.0. But 5.1.1 seems ok.

It crashes with 5.1.1 btw. (the Archlinux binary). I'm now going to try the `unison-upstream-bin` AUR package.

Although I probably won't be able to test a lot more today.

**Comment by tleedjarv** [comment]
@FSMaxB thank you for testing.

What is interesting is that I have not been able to crash a remote sync, even if I limit the stack size to the smallest possible with a GTK application. Remote syncs do not show any signs of stack exhaustion.

I have finally been able to make a local sync crash and can now investigate further.

**Comment by tleedjarv** [comment]
> It crashes with 5.1.1 btw. (the Archlinux binary). I'm now going to try the `unison-upstream-bin` AUR package.

That is weird, the unison binary I get from Arch claims it's been compiled with 5.1.0. Could this be a difference in Arch mirrors? (I know nothing about Arch, so there's that...)

I'd like to test with Arch 5.1.1 binary if you can instruct me how I can get it.

**Comment by FSMaxB** [comment]
> That is weird, the unison binary I get from Arch claims it's been compiled with 5.1.0. Could this be a difference in Arch mirrors? (I know nothing about Arch, so there's that...)

Oh, sorry about that. I totally forgot that this was still the binary I produced to try to get more debug information.

I took the archlinux PKGBUILD, removed the patch and just rebuilt it locally. https://gitlab.archlinux.org/archlinux/packaging/packages/unison This is then automatically using OCaml 5.1.1.

BUT: I can confirm that the crashes don't happen with `unison-upstream-bin`.

**Comment by FSMaxB** [comment]
Last test for today, the binary from this link: https://github.com/tleedjarv/unison/actions/runs/8481008964

This does not crash. So it seems like the crashes, for me at least are actually only happening with Archlinux produced binaries :eyes:

**Comment by tleedjarv** [comment]
Thank you @FSMaxB. Your testing confirms my findings that Arch-built binaries do seem to behave differently.

**Comment by tleedjarv** [comment]
I have a patch that I hope will resolve the crashes that I've been able to reproduce (and hopefully those seen by each reporter).

@callegar @FSMaxB if you are building from source then please apply this patch: https://github.com/tleedjarv/unison/commit/b18679153663d7cebfc55b73cb86eebefb6a9862
I would really appreciate it if you could test a binary built on archlinux with the patch applied. Since archlinux binaries have been behaving worse then this would be the primary testing subject.

Alternatively, you can also test a CI-built binary that you can download from https://github.com/tleedjarv/unison/actions/runs/8498976286 (that's a clean 2.53.4 with only this patch applied).

If you have a hard time reproducing the crash then try to limit the stack size (for me 144 KB for the GUI worked, for example). That should limit the number of updates required to trigger the crash. You can also artificially create a large number of updates to sync (just empty files, for example).



Clarifications on some previous comments:


> I took the archlinux PKGBUILD, removed the patch and just rebuilt it locally. https://gitlab.archlinux.org/archlinux/packaging/packages/unison This is then automatically using OCaml 5.1.1.

Yes, the patch in PKGBUILD (largefiles) is bogus and should be removed. Also, PKGBUILD lists `emacs` as a build dep, but it's really not.



> > The takeaway is that don't compile with OCaml 5.1.0. But 5.1.1 seems ok.
> 
> It crashes with 5.1.1 btw. ~(the Archlinux binary)~ (Correction: A binary built using the Archlinux PKGBUILD, but with OCaml 5.1.1 instead of 5.1.0 and without the archlinux patch).

Ok, so we know it still crashes with 5.1.1 but the binary built with 5.1.1 is _much_ faster. Let's see if the patch fixes the crash.

**Comment by Octachron** [comment]
@tleedjarv , looking at your comment on the stack size, are you limiting the OCaml stack size with `OCAMLRUNPARAM=l=...`? OCaml 5 has a separate OCaml stack (to implement a segmented stack for effect handlers) and using `ulimit` will not affect the OCaml-managed call stack.

**Comment by tleedjarv** [comment]
> @tleedjarv , looking at your comment on the stack size, are you limiting the OCaml stack size with `OCAMLRUNPARAM=l=...`? OCaml 5 has a separate OCaml stack (to implement a segmented stack for effect handlers) and using `ulimit` will not affect the OCaml-managed call stack.

Ah, that explains the results I got! Thank you for pointing it out.
Is the OCaml manual out of date then? It still reads as follows:
> (stack_limit) The limit (in words) of the stack size. This is only relevant to the byte-code runtime, as the native code runtime uses the operating system’s stack. 

I need to retest with `OCAMLRUNPARAM=l=...` but I do believe I've already found the issue in Unison code.

**Comment by tleedjarv** [comment]
I have now tested the tentative fix with `OCAMLRUNPARAM=l=1000` and 40 000 updates to sync. Confirmed, this is a fix to _a_ problem. I'd like to receive some feedback from Arch users before confirming it as the complete fix to this ticket.

**Comment by callegar** [comment]
Unfortunately, I am under a deadline and I will not be able to test before next week. Hope someone is faster than me. Thanks for all the analysis and fix work!

**Comment by tleedjarv** [comment]
I propose this ticket can be closed once #1016 is merged even if all testing on Arch is not finished. If the crashes persist then it can be reopened.

In conclusion:
* I was wrong in suggesting Arch is the cause of this issue.
* The underlying issue appears to be a non-tail call recursion. Yet, it's not as clear cut, as apparently it does not cause stack overflow under all conditions. The code has been present since 2.53.1 (released a year ago).
  * I could not reproduce the crash with the same conditions as some of the reporters; but I could (finally) reproduce it under different conditions.
* That said, there is still something weird going on with the binary from Arch.
  * It could just be the extreme slowness (even bytecode builds are faster!), if nothing else.
  * Based on the reports it seems that the stack overflow crashes were more common on Arch; there is just one other report in #990.
  * The crash, at least in my testing, is different with the binary from Arch. I don't think this requires further investigation unless the problems persist even with the patch.

**Comment by alerque** [comment]
Thanks for looking into this. I've dropped the patch (originally from Fedora) that Arch Linux was carrying to fix problems you claim are now obsolete. That rebuild is now in the official [extra] repo.

If there are ongoing issues with performance could somebody please open up a new issue either here (and ping me) or on the [Arch Linux packaging repo](https://gitlab.archlinux.org/archlinux/packaging/packages/unison/-/issues) specifically about that? If we're doing something wrong with build flags and optimizations I'd be happy to get it fixed. I'm not very familiar with OCAML but can certainly help with getting things changed if anybody knows what to do. If build logs showing various flags or whatever would be helpful ask in an issue and I'll see what I can provide. Also all the tooling for building reproducible copies of Arch Linux packages and then playing with your own build settings is all open source and pretty easy to setup. I can help somebody with that if you want to try to replicate build issues.

**Comment by tleedjarv** [comment]
I had a quick look at the PKGBUILD and I don't see anything there that could have caused the performance issue. It may have been due to 5.1.0 compiler having some regression. If I understand correctly, the new builds are compiled with 5.1.1 (my own build on Arch with 5.1.1 did not seem to suffer from performance issues). I had suspected that this may have been a bytecode build, but at least in my testing, it was even slower than a bytecode build! @alerque is there a way to see the build logs somewhere?

@alerque A side remark about PKGBUILD: you can remove emacs as build dependency, you can remove `DEBUGGING=false THREADS=true` make vars (no longer used) and you can also remove `export CFLAGS=`, unless the build starts failing (it used to fail with CFLAGS set but this has been fixed). But none of this should have impacted the build in a negative way.

For now, I would like to think that this was some regression with the 5.1.0 compiler and doesn't require any further attention, but I will also test the new binary from the rebuild you mentioned.

**Comment by Octachron** [comment]
There was an important GC performance regression between 5.0.0 and 5.1.0 which was fixed in 5.1.1. The regression was due to a bad interaction between the GC and bigarrays or C-allocated objects, I don't know if this applies to unison, but this is at least a likely culprit.

**Comment by tleedjarv** [comment]
> bigarrays or C-allocated objects, I don't know if this applies to unison

Yes, it does apply. This surely must have been it.

**Comment by tleedjarv** [comment]
I tested pkgrel 2, built with 5.1.0, slow. Now I tested pkgrel 4, built with 5.1.1, performance is ok. I think this concludes the performance issue, so we no longer need the build logs.
